bpf/init.sh

```diff
@@ -296,16 +309,49 @@ function bpf_clear_cgroups()
 	HOOK=$2
 	NAME=$3
 
+	# Since Linux commit 1ba5ad36e00f ("bpftool: Use libbpf_bpf_attach_type_str"),
+	# bpftool uses the libbpf_bpf_attach_type_str() format in 'bpftool cgroup
+	# show' output. Perform a naive translation to ensure compatibility with prior
+	# bpftool versions and to avoid updating the hook name at all call sites. The
+	# transformed string can be used in a suffix match against the new format.
+
+	# Examples: (old -> transformed, new)
+	# connect4 -> 4_connect, cgroup_inet4_connect
+	# post_bind6 -> 6_post_bind, cgroup_inet6_post_bind
+	# sendmsg4 -> 4_sendmsg, cgroup_udp4_sendmsg
+
+	# There is no inet4/6 variant of e.g. sendmsg, so it's safe to assume the
+	# intended udp4/6 hook is selected.
+	newhook=$(echo "$HOOK" | sed -E 's/([a-z_]+)([0-9])/\2_\1/')
+
 	set +e
 	bpftool link detach pinned "$BPFFS_ROOT/cilium/socketlb/links/cgroup/$NAME" || true
 	rm -f "$BPFFS_ROOT/cilium/socketlb/links/cgroup/$NAME"
-
-	ID=$(bpftool cgroup show $CGRP | grep -w $HOOK | awk '{print $1}')
 	set -e
-	
-	if [ -n "$ID" ]; then
-		bpftool cgroup detach $CGRP $HOOK id $ID
-	fi
+
+	# Get all programs attached to the given cgroup and store their ids in a
+	# space-separated string. Run jq in raw output mode (-r) to prevent the
+	# resulting string from being written to stdout with double quotes. Perform a
+	# full match on the 'legacy' hook name appearing in older versions of bpftool,
+	# but perform a suffix match using the 'new' hook name.
+	ids=$(bpftool cgroup show "$CGRP" -j |
+		jq -r --arg legacy "$HOOK" --arg new "$newhook" '. |
+			map(. |
+				select(
+					.attach_type == $legacy or
+					(.attach_type | endswith($new))
+				)
+				| .id)
+			| join(" ")')
```

Comment:
I think this can be simplified, although I haven't tested in situation:

```suggestion
		jq -r --arg legacy "$HOOK" --arg new "$newhook" '.[] |
			select(
				.attach_type == $legacy or
				(.attach_type | endswith($new))
			) | .id)')
```

The output won't be on a single line, but should work all the same for the `for` loop below

Reply:
> but should work all the same for the `for` loop below

iirc this is not the case, `for` loops over words by default. Looping over lines with `for` can be done by setting `IFS=$'\n'`, but I'd rather make jq output the right format.

Reply:
Looping on lines seems to work well when trying locally. But OK, fine by me anyway.
You can still remove the first `. |` and start directly with the `map(`, but that's really a nitpick.

Reply:
Sorry, you're right. Default IFS seems to be set to `<space><tab><newline>`. I assume that's been the case for a long time, not sure why I thought it was different. I'll go with your suggestion. :+1: